### PR TITLE
add QWen

### PR DESCRIPTION
**Pull request by wtmlon**
### Description

新增通义千问模型
```python
import paddle
from paddlenlp.transformers.qwen.modeling import QWenForCausalLM
from paddlenlp.transformers.qwen.tokenizer import QWenTokenizer
from paddlenlp.transformers.qwen.configuration import QWenConfig

tokenizer = QWenTokenizer.from_pretrained('Qwen/Qwen-7B')
inputs = tokenizer('蒙古国的首都是乌兰巴托（Ulaanbaatar）\n冰岛的首都是雷克雅未克（Reykjavik）\n埃塞俄比亚的首都是', return_tensors='pd', padding=True)

model = QWenForCausalLM.from_pretrained('Qwen/Qwen-7B')
output = model.generate(input_ids=inputs['input_ids'], attention_mask=inputs['attention_mask'], decode_strategy='greedy_search', max_length=100)
print(tokenizer.decode(output[0][0])

'''
'亚的斯亚贝巴（Addis Ababa）\n尼日利亚的首都是阿布贾（Abuja）\n巴基斯坦的首都是伊斯兰堡（Islamabad）\n菲律宾的首都是马尼拉（Manila）\n波兰的首都是华沙（Warsaw）\n葡萄牙的首都是里斯本（Lisbon）\n俄罗斯的首都是莫斯科（Moscow）\n新加坡的首都是新加坡（Singapore）\n南非的首都是比勒陀利亚（Pretoria）\n西班牙的首'
'''
```

### CODE REVIEW

File: paddlenlp/transformers/qwen/configuration.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        emb_dropout_prob=0.0,
+        attn_dropout_prob=0.0,
+        layer_norm_epsilon=1e-6,
+        initializer_range=0.02,
+        max_position_embeddings=8192,
+        scale_attn_weights=True,
+        use_cache=True,
+        bf16=False,
+        fp16=False,
+        fp32=False,
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/configuration.py]: 这里可以仿照源代码用这几个输入，但是最后不要定义self.bf16这几个，避免和其他模型不一样，加入逻辑统一至self.dtype

File: paddlenlp/transformers/qwen/modeling.py [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+        assert self.projection_size % config.num_attention_heads == 0
+        self.hidden_size_per_attention_head = self.projection_size // config.num_attention_heads
+
+        self.c_attn = nn.Linear(config.hidden_size, 3 * self.projection_size)
+
+        self.c_proj = nn.Linear(config.hidden_size, self.projection_size, bias_attr=not config.no_bias)
+
+        self.is_fp32 = not (config.bf16 or config.fp16)
+
+        self.bf16 = config.bf16
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 统一通过config里的dtype来判定吧

File: paddlenlp/transformers/qwen/modeling.py [hunk 3]
```diff
@@ -0,0 +1,806 @@
...
+        else:
+            assert config.rotary_pct < 1
+            self.rotary_ndims = int(self.hidden_size_per_attention_head * config.rotary_pct)
+        dim = self.rotary_ndims if self.rotary_ndims is not None else self.hidden_size_per_attention_head
+        self.rotary_emb = RotaryEmbedding(dim, base=config.rotary_emb_base)
+
+        self.use_dynamic_ntk = config.use_dynamic_ntk
+        self.use_logn_attn = config.use_logn_attn
+
+        logn_list = [math.log(i, self.seq_length) if i > self.seq_length else 1 for i in range(1, 32768)]
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: no magic number, 将32768这个数定义成一个变量, 比如max_ntk_seq_length

File: paddlenlp/transformers/qwen/modeling.py [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+        attn_weights = paddle.matmul(query, key.transpose([0, 1, 3, 2]))
+
+        if self.scale_attn_weights:
+            attn_weights = attn_weights * self.inv_norm_factor
+
+        query_length, key_length = query.shape[-2], key.shape[-2]
+        causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length]
+        mask_value = paddle.finfo(attn_weights.dtype).min
+        mask_value = paddle.full_like(attn_weights, mask_value, dtype=attn_weights.dtype)
+        attn_weights = paddle.where(causal_mask, attn_weights, mask_value)
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 外面准备好 -inf/0 的float mask, 这里的逻辑改成+

File: paddlenlp/transformers/qwen/modeling.py [hunk 5]
```diff
@@ -0,0 +1,806 @@
...
+
+        attn_weights = nn.functional.softmax(attn_weights, axis=-1)
+
+        attn_weights = attn_weights.astype(value.dtype)
+        attn_weights = self.attn_dropout(attn_weights)
+
+        if head_mask is not None:
+            attn_weights = attn_weights * head_mask
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 模型有用到head_mask吗？我们这边模型基本都没有适配head_mask

File: paddlenlp/transformers/qwen/modeling.py [hunk 6]
```diff
@@ -0,0 +1,806 @@
...
+
+        return outputs
+
+
+class QWenPreTrainedModel(PretrainedModel):
+    config_class = QWenConfig
+    base_model_prefix = "transformer"
+    is_parallelizable = False
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["QWenBlock"]
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 检查一下, 我们应该没有这几个attributes

File: paddlenlp/transformers/qwen/modeling.py [hunk 7]
```diff
@@ -0,0 +1,806 @@
...
+        elif inputs_embeds is not None:
+            input_shape = inputs_embeds.shape[:-1]
+            batch_size = inputs_embeds.shape[0]
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        # device = input_ids.device if input_ids is not None else inputs_embeds.device
+
+        if token_type_ids is not None:
+            token_type_ids = token_type_ids.reshape([-1, input_shape[-1]])
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 确认一下这个模型是不是真的有token_type_ids

File: paddlenlp/transformers/qwen/modeling.py [hunk 8]
```diff
@@ -0,0 +1,806 @@
...
+
+        if position_ids is None:
+            position_ids = paddle.arange(
+                past_length,
+                input_shape[-1] + past_length,
+                dtype=paddle.int64,
+            )
+            position_ids = position_ids.unsqueeze(0).reshape([-1, input_shape[-1]])
+
+        if attention_mask is not None:
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 这块的逻辑和其他模型统一

File: paddlenlp/transformers/qwen/modeling.py [hunk 9]
```diff
@@ -0,0 +1,806 @@
...
+            if batch_size <= 0:
+                raise ValueError("batch_size has to be defined and > 0")
+            attention_mask = attention_mask.reshape([batch_size, -1])
+            attention_mask = attention_mask[:, None, None, :]
+            attention_mask = attention_mask.astype(paddle.bfloat16)
+            attention_mask = (1.0 - attention_mask) * paddle.finfo(paddle.bfloat16).min
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 不能hardcode bf16

File: paddlenlp/transformers/qwen/modeling.py [hunk 10]
```diff
@@ -0,0 +1,806 @@
...
+
+        encoder_attention_mask = None
+        head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 看下有没有用到，没有就不用支持

File: paddlenlp/transformers/qwen/modeling.py [hunk 11]
```diff
@@ -0,0 +1,806 @@
...
+
+
+class QWenLMHeadModel(QWenPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [r"h\.\d+\.attn\.rotary_emb\.inv_freq"]
+    _keys_to_ignore_on_load_unexpected = [r"h\.\d+\.attn\.masked_bias"]
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.transformer = QWenModel(config)
+        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias_attr=False)
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 后续检查一下这个写法能不能跑通sharding stage 3

File: llm/data.py [hunk 12]
```diff
@@ -26,7 +26,7 @@ def get_convert_example(model):
 
     if base_model_prefix == "chatglm":
         return convert_example_chatglm
-    elif base_model_prefix in ["chatglm_v2", "llama", "bloom", "opt"]:
+    elif base_model_prefix in ["chatglm_v2", "llama", "bloom", "opt", "transformer"]:
```
Comment (sijunhe) [inline llm/data.py]: 把这个base_model_prefix给换了吧，没有辨识度

File: llm/finetune_generation.py [hunk 13]
```diff
@@ -112,6 +113,10 @@ def main():
     if isinstance(tokenizer, LlamaTokenizer):
         tokenizer.pad_token = tokenizer.eos_token if tokenizer.eos_token else "<pad>"
 
+    if isinstance(tokenizer, QWenTokenizer):
+        tokenizer.pad_token_id = 151643
+        tokenizer.eos_token_id = 151643
```
Comment (sijunhe) [inline llm/finetune_generation.py]: 直接在tokenizer里面定义可以吗？

File: paddlenlp/transformers/qwen/tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+    def __init__(
+        self,
+        vocab_file,
+        errors="replace",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        if not is_tiktoken_available():
+            raise ValueError("tiktoken is not installed, please install it use: pip install tiktoken")
+        global tiktoken
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/tokenizer.py]: 为啥这里要global一下?

File: paddlenlp/transformers/qwen/modeling.py [hunk 15]
```diff
@@ -0,0 +1,910 @@
...
+    else:
+        logits = paddle.matmul(x, y, transpose_y=False)
+        return logits
+
+
+class QWenAttention(nn.Layer):
+    def __init__(self, config):
+        super().__init__()
+
+        self.register_buffer("masked_bias", paddle.to_tensor(-1e4), persistable=False)
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 这个masked_bias用在哪里了？好像没看到

File: paddlenlp/transformers/qwen/modeling.py [hunk 16]
```diff
@@ -0,0 +1,910 @@
...
+            outputs = (hidden_states,) + outputs[1:]
+
+        return outputs
+
+
+class QWenPreTrainedModel(PretrainedModel):
+    config_class = QWenConfig
+    base_model_prefix = "qwen"
+    # will use in unittest
+    supports_gradient_checkpointing = True
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 这个不需要吧？

File: paddlenlp/transformers/qwen/modeling.py [hunk 17]
```diff
@@ -0,0 +1,910 @@
...
+                    paddle.tensor.normal(
+                        mean=0.0,
+                        std=self.config.initializer_range / math.sqrt(2 * self.config.num_hidden_layers),
+                        shape=p.shape,
+                    )
+                )
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, QWenModel):
+            module.enable_recompute = value
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: 不需要?

File: paddlenlp/transformers/qwen/modeling.py [hunk 18]
```diff
@@ -0,0 +1,910 @@
...
+            encoder_attention_mask,
+            use_cache,
+            output_attentions,
+            use_reentrant=False,
+        )
+        return hidden_states
+
+    def get_masks(self, batch_size, seq_length, past_length, padding_mask=None):
+        # casual mask
+        casual_mask = paddle.tril(paddle.ones([batch_size, 1, seq_length, seq_length])).astype("bool")
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]:
```suggestion
        casual_mask = paddle.tril(paddle.ones([batch_size, 1, seq_length, seq_length], dtype="bool"))
```

File: paddlenlp/transformers/qwen/modeling.py [hunk 19]
```diff
@@ -0,0 +1,910 @@
...
+
+        hidden_states = self.drop(hidden_states)
+        output_shape = input_shape + [
+            hidden_states.shape[-1],
+        ]
+
+        if self.enable_recompute and self.training:
+            if use_cache:
+                logger.warning_once(
+                    "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: `use_cache=True` is incompatible with `recompute`

File: paddlenlp/transformers/qwen/modeling.py [hunk 20]
```diff
@@ -0,0 +1,910 @@
...
+        hidden_states = transformer_outputs[0]
+
+        lm_logits = self.lm_head(hidden_states)
+
+        loss = None
+        if labels is not None:
+            # TODO shift_label if
+            # labels = labels.to(lm_logits.device)
+            shift_logits = lm_logits[..., :-1, :]
+            shift_labels = labels[..., 1:]
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: rm

File: paddlenlp/transformers/qwen/modeling.py [hunk 21]
```diff
@@ -0,0 +1,910 @@
...
+            loss_fct = nn.CrossEntropyLoss()
+            loss = loss_fct(shift_logits.reshape([-1, shift_logits.shape[-1]]), shift_labels.reshape([-1]))
```
Comment (sijunhe) [inline paddlenlp/transformers/qwen/modeling.py]: no need to reshape

**Review by sijunhe — APPROVED** [review]
